File: sources/CSharp/src/Ers/SubModel/Component/PathComponent.cs
using System;
using System.Numerics;
using System.Runtime.InteropServices;
using Ers.Engine;

namespace Ers
{
    /// <summary>
    /// Component to define paths along which entities can be rendered and animated.
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    public struct PathComponent : ICoreComponent
    {
        /// <summary>
        /// The number of segments in the path.
        /// </summary>
        public int NumSegments
        {
            get => ErsEngine.ERS_PathComponent_GetNumSegments(CorePtr);
        }

        public PathSegment GetSegment(int index) => new PathSegment(ErsEngine.ERS_PathComponent_GetSegment(CorePtr, index));

        public void AddStraight(Vector3 from, Vector3 to)
        {
            ErsEngine.ERS_PathComponent_AddStraight(CorePtr, from.X, from.Y, from.Z, to.X, to.Y, to.Z);
        }
        public void AddHelical(Vector3 center, float radius, float beginAngle, float endAngle, float endZ)
        {
            ErsEngine.ERS_PathComponent_AddHelical(CorePtr, center.X, center.Y, center.Z, radius, beginAngle, endAngle, endZ);
        }

        public void AddCubicBezier(Vector3 p0, Vector3 p1, Vector3 p2, Vector3 p3)
        {
            ErsEngine.ERS_PathComponent_AddCubicBezier(CorePtr, p0.X, p0.Y, p0.Z, p1.X, p1.Y, p1.Z, p2.X, p2.Y, p2.Z, p3.X, p3.Y, p3.Z);
        }

        public void AddCubicBezierFromDirections(Vector3 start, Vector3 startDir, Vector3 end, Vector3 endDir, float curvature)
        {
            ErsEngine.ERS_PathComponent_AddCubicBezierFromDirections(
                CorePtr, start.X, start.Y, start.Z, startDir.X, startDir.Y, startDir.Z, end.X, end.Y, end.Z, endDir.X, endDir.Y, endDir.Z,
                curvature);
        }
        public static nuint CoreTypeId() => ErsEngine.ERS_PathComponent_TypeId();

        /// <summary>
        /// Native pointer to the core instance.
        /// </summary>
        public IntPtr CorePtr
        {
            get {
                unsafe
                {
                    fixed(PathComponent* ptr = &this)
                    {
                        return (IntPtr)ptr;
                    }
                }
            }
        }
    }
}

File: sources/CPP/include/Ers/SubModel/Component/PathComponent.cpp
#include "PathComponent.h"
#include "Ers/Api.h"
#include "Ers/Math/PathSegment.h"

namespace Ers
{
    uint32_t PathComponent::CoreTypeId()
    {
        return Ers::Engine::ERS_PathComponent_TypeId();
    }

    int PathComponent::GetNumSegments() const
    {
        return Ers::Engine::ERS_PathComponent_GetNumSegments(const_cast<PathComponent*>(this));
    }

    PathSegment* PathComponent::GetSegment(int index)
    {
        void* ptr = Ers::Engine::ERS_PathComponent_GetSegment(this, index);
        return static_cast<PathSegment*>(ptr);
    }

    void PathComponent::AddStraight(Vector3 from, Vector3 to)
    {
        Ers::Engine::ERS_PathComponent_AddStraight(this, from.X, from.Y, from.Z, to.X, to.Y, to.Z);
    }

    void PathComponent::AddHelical(Vector3 center, float radius, float beginAngle, float endAngle, float endZ)
    {
        Ers::Engine::ERS_PathComponent_AddHelical(this, center.X, center.Y, center.Z, radius, beginAngle, endAngle, endZ);
    }

    void PathComponent::AddCubicBezier(Vector3 p0, Vector3 p1, Vector3 p2, Vector3 p3)
    {
        Ers::Engine::ERS_PathComponent_AddCubicBezier(this, p0.X, p0.Y, p0.Z, p1.X, p1.Y, p1.Z, p2.X, p2.Y, p2.Z, p3.X, p3.Y, p3.Z);
    }

    void PathComponent::AddCubicBezierFromDirections(Vector3 start, Vector3 startDir, Vector3 end, Vector3 endDir, float curvature)
    {
        Ers::Engine::ERS_PathComponent_AddCubicBezierFromDirections(
            this, start.X, start.Y, start.Z, startDir.X, startDir.Y, startDir.Z, end.X, end.Y, end.Z, endDir.X, endDir.Y, endDir.Z,
            curvature);
    }
} // namespace Ers
